### PR TITLE
WIP: Alternate Infrastructure Provider for AWS

### PR DESCRIPTION
**Pull request by patrickdillon**
Depends on #7488 
(it would be best to contain reviews of the first four commits within that PR)

The focus of this PR is introducing a build tag to allow building binaries with a separate dependency path--allowing us to create binaries that include/exclude dependencies such as terraform based on the tag. 

- use a `altinfra` build flag to build installer without terraform

```shell
[padillon@fedora installer]$ TAGS=altinfra ./hack/build.sh &> /dev/null
[padillon@fedora installer]$ ls -sh bin/openshift-install 
216M bin/openshift-install
[padillon@fedora installer]$ ./hack/build.sh &> /dev/null
[padillon@fedora installer]$ ls -sh bin/openshift-install 
620M bin/openshift-install
```

I also checked with `-v` when running `go build` to check that terraform packages were built (or not) as expected with the build flag. The dog ate my homework on that one, and for some (sort of caching?) reason the output now shows nothing. Will need to debug once I have more time. 

This PR will also introduce an alternative AWS infrastructure provider for demo purposes; although we will probably want to ultimately merge those separately in different PRs.

### CODE REVIEW

**Comment by patrickdillon** [comment]
> /cc @andfasano @r4f4

oops I meant to request reviews on https://github.com/openshift/release/pull/43826
which enables testing of this

File: pkg/infrastructure/aws/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+	// We sleep here otherwise got an error when creating the ec2 instance referencing the profile.
+	time.Sleep(10 * time.Second)
+	return instanceProfile.Arn, nil
+}
+
+func existingRole(client iamiface.IAMAPI, roleName string) (*iam.Role, error) {
+	result, err := client.GetRole(&iam.GetRoleInput{RoleName: aws.String(roleName)})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() == iam.ErrCodeNoSuchEntityException {
```
Comment (r4f4) [inline pkg/infrastructure/aws/bootstrap.go:272]: See pattern above.

File: pkg/infrastructure/aws/vpc.go [hunk 2]
```diff
@@ -0,0 +1,1362 @@
...
+	}
+
+	const duplicatePermissionErrorCode = "InvalidPermission.Duplicate"
+	if len(egressToAuthorize) > 0 {
+		err = retry.OnError(backoff, func(error) bool { return true }, func() error {
+			_, err := client.AuthorizeSecurityGroupEgress(&ec2.AuthorizeSecurityGroupEgressInput{
+				GroupId:       aws.String(securityGroupID),
+				IpPermissions: egressToAuthorize,
+			})
+			var awsErr awserr.Error
```
Comment (r4f4) [inline pkg/infrastructure/aws/vpc.go:385]: I'm pretty sure this needs to be a pointer:
```suggestion
			var awsErr *awserr.Error
```

File: pkg/infrastructure/aws/vpc.go [hunk 3]
```diff
@@ -0,0 +1,1362 @@
...
+			Tags:      append(ec2Tags(o.InfraID, fmt.Sprintf("%s-eip-%s", o.InfraID, availabilityZone)), o.additionalEC2Tags...),
+		})
+		return err
+	})
+	if err != nil {
+		return "", fmt.Errorf("cannot tag NAT gateway EIP: %w", err)
+	}
+
+	isNATGatewayRetriable := func(err error) bool {
+		if awsErr, ok := err.(awserr.Error); ok {
```
Comment (r4f4) [inline pkg/infrastructure/aws/vpc.go:617]: Use pattern above.

File: pkg/infrastructure/aws/vpc.go [hunk 4]
```diff
@@ -0,0 +1,1362 @@
...
+	internalLBName := fmt.Sprintf("%s-int", o.InfraID)
+
+	// Check if the internal load balancer already exists.
+	describeLBInput := &elbv2.DescribeLoadBalancersInput{
+		Names: []*string{&internalLBName},
+	}
+	describeLBOutput, err := elbClient.DescribeLoadBalancers(describeLBInput)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() != "LoadBalancerNotFound" {
```
Comment (r4f4) [inline pkg/infrastructure/aws/vpc.go:865]: use pattern above

File: pkg/infrastructure/aws/vpc.go [hunk 5]
```diff
@@ -0,0 +1,1362 @@
...
+	internalATGName := fmt.Sprintf("%s-aint", o.InfraID)
+
+	// Check if the target group already exists
+	describeInternalATGInput := &elbv2.DescribeTargetGroupsInput{
+		Names: []*string{&internalATGName},
+	}
+	describeInternalATGOutput, err := elbClient.DescribeTargetGroups(describeInternalATGInput)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() != "TargetGroupNotFound" {
```
Comment (r4f4) [inline pkg/infrastructure/aws/vpc.go:926]: Use pattern above

File: pkg/infrastructure/aws/vpc.go [hunk 6]
```diff
@@ -0,0 +1,1362 @@
...
+	// Create InternalS TargetGroup.
+	internalSTGName := fmt.Sprintf("%s-sint", o.InfraID)
+	// Check if the target group already exists
+	describeInternalSTGInput := &elbv2.DescribeTargetGroupsInput{
+		Names: []*string{&internalSTGName},
+	}
+	describeInternalSTGOutput, err := elbClient.DescribeTargetGroups(describeInternalSTGInput)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() != "TargetGroupNotFound" {
```
Comment (r4f4) [inline pkg/infrastructure/aws/vpc.go:1015]: Use pattern above

File: pkg/infrastructure/aws/vpc.go [hunk 7]
```diff
@@ -0,0 +1,1362 @@
...
+	// Create external LB.
+	externalLBName := fmt.Sprintf("%s-ext", o.InfraID)
+	// Check if the internal load balancer already exists
+	describeExternalLBInput := &elbv2.DescribeLoadBalancersInput{
+		Names: []*string{&externalLBName},
+	}
+	describeExternalLBOutput, err := elbClient.DescribeLoadBalancers(describeExternalLBInput)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() != "LoadBalancerNotFound" {
```
Comment (r4f4) [inline pkg/infrastructure/aws/vpc.go:1112]: Use pattern above

File: pkg/infrastructure/aws/vpc.go [hunk 8]
```diff
@@ -0,0 +1,1362 @@
...
+	externalTGName := fmt.Sprintf("%s-aext", o.InfraID)
+
+	// Check if the target group already exists
+	describeExternalTGInput := &elbv2.DescribeTargetGroupsInput{
+		Names: []*string{&externalTGName},
+	}
+	describeExternalTGOutput, err := elbClient.DescribeTargetGroups(describeExternalTGInput)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() != "TargetGroupNotFound" {
```
Comment (r4f4) [inline pkg/infrastructure/aws/vpc.go:1173]: use pattern above

File: pkg/terraform/stages/azure/stages.go [hunk 9]
```diff
@@ -1,48 +1,95 @@
...
 
 import (
+	"os"
+
+	"github.com/openshift/installer/pkg/infrastructure"
 	"github.com/openshift/installer/pkg/terraform"
 	"github.com/openshift/installer/pkg/terraform/providers"
 	"github.com/openshift/installer/pkg/terraform/stages"
 	typesazure "github.com/openshift/installer/pkg/types/azure"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/terraform/stages/azure/stages.go:11]: Archived package. See comment above.

File: pkg/terraform/stages/ibmcloud/stages.go [hunk 10]
```diff
@@ -1,27 +1,53 @@
 package ibmcloud
 
 import (
+	"os"
+
+	"github.com/openshift/installer/pkg/infrastructure"
 	"github.com/openshift/installer/pkg/terraform"
 	"github.com/openshift/installer/pkg/terraform/providers"
 	"github.com/openshift/installer/pkg/terraform/stages"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/terraform/stages/ibmcloud/stages.go:10]: See above

File: pkg/terraform/stages/libvirt/stages.go [hunk 11]
```diff
@@ -1,22 +1,47 @@
 package libvirt
 
 import (
+	"os"
+
+	"github.com/openshift/installer/pkg/infrastructure"
 	"github.com/openshift/installer/pkg/terraform"
 	"github.com/openshift/installer/pkg/terraform/providers"
 	"github.com/openshift/installer/pkg/terraform/stages"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/terraform/stages/libvirt/stages.go:10]: See above

File: pkg/terraform/stages/nutanix/stages.go [hunk 12]
```diff
@@ -1,13 +1,34 @@
 package nutanix
 
 import (
+	"os"
+
+	"github.com/openshift/installer/pkg/infrastructure"
 	"github.com/openshift/installer/pkg/terraform"
 	"github.com/openshift/installer/pkg/terraform/providers"
 	"github.com/openshift/installer/pkg/terraform/stages"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/terraform/stages/nutanix/stages.go:10]: See above

File: pkg/terraform/stages/openstack/stages.go [hunk 13]
```diff
@@ -1,23 +1,48 @@
 package openstack
 
 import (
+	"os"
+
+	"github.com/openshift/installer/pkg/infrastructure"
 	"github.com/openshift/installer/pkg/terraform"
 	"github.com/openshift/installer/pkg/terraform/providers"
 	"github.com/openshift/installer/pkg/terraform/stages"
+	"github.com/pkg/errors"
```
Comment (r4f4) [inline pkg/terraform/stages/openstack/stages.go:10]: See above.

**Comment by patrickdillon** [comment]
/hold
to stop tests from running unnecessarily

**Pull request closed without merging**